ow_power_system/src/power_system_node.cpp

```diff
@@ -13,92 +13,116 @@
-  bool csv_path_param_exist = nh.param("power_draw_csv_path", csv_path,
-    ros::package::getPath("ow_power_system") + default_csv);
-
-  if (!csv_path_param_exist)
-  {
-    ROS_WARN_NAMED("power_system_node", "power_draw_csv_path param was not set! Using default value: data_const_load.csv");
-  }
-
-  ROS_INFO_STREAM_NAMED("power_system_node", "power_draw_csv_path is set to: " << csv_path);
+float update_count = 0.00; // Timestep initialization
+float battery_lifetime = 2738.0; // Estimate of battery lifetime (seconds)
+float init_power = 0.00; // Power initialization
+float init_temp = 20.00; // Temperature initialization
+float init_voltage = 4.10; // Voltage initialization
 
-  // Read battery data from a file.
-  auto power_data = read_file(csv_path);
-
-  // Create a configuration from a file
-  string config_path = ros::package::getPath("ow_power_system")+"/config/example.cfg";
-  ConfigMap config(config_path);
-  // Contruct a new prognoser using the prognoser factory. The prognoser
-  // will automatically construct an appropriate model, observer and predictor
-  // based on the values specified in the config.
-  auto prognoser = PrognoserFactory::instance().Create("ModelBasedPrognoser", config);
+class SubscribeAndPublish
+{
+public:
+  SubscribeAndPublish()
+  {
+    //Construct our State of Charge (SOC) publisher
+    SOC_pub = nh.advertise<std_msgs::Float64>("power_system_node/state_of_charge",1000);
+    //Construct our Remaining Useful Life (RUL) publisher (Seconds)
+    RUL_pub = nh.advertise<std_msgs::Int16>("power_system_node/remaining_useful_life",1000);
+    //Construct our Battery Temperature (TempBat) publisher
+    TempBat_pub = nh.advertise<std_msgs::Float64>("power_system_node/battery_temperature",1000);
 
-  // Retrieve our publication rate expressed in Hz
-  double power_update_rate = 1;
-  double power_update_rate_override;  // allow the user to override it
-  bool update_rate_param_exist = nh.param("power_update_rate", power_update_rate_override, 0.1);
+    //Construct our Average Mechanical Power subsriber (Watts)
+    MechPower_sub = nh.subscribe("/mechanical_power/average", 1000, &SubscribeAndPublish::powerCallback, this);
 
-  if (update_rate_param_exist)
+  }
+  void powerCallback(const std_msgs::Float64::ConstPtr& msg)
   {
-    ROS_WARN_NAMED("power_system_node", "Overriding the default power_update_rate!");
-    // Validated the parameter
-    if (power_update_rate == 0)
+    using namespace std::chrono;
+    ros::NodeHandle nh;
+
+    // Set mechanical power value to rostopic subscription
+    float mech_power = msg->data; // [W]
+
+    // Create temperature estimate with pseudorandom noise generator
+    float min_temp = 17.5; // minimum temp = 17.5 deg. C
+    float max_temp = 21.5; // maximum temp = 21.5 deg. C
+
+    // Temperature estimate based on pseudorandom noise and fixed range
+    float temp_est = min_temp + static_cast <float> (rand()) / (static_cast <float> (RAND_MAX/(max_temp - min_temp)));
```

Comment:
I don't know where you got this formula to perform range random but it wouldn't give you the desired results
```suggestion
    float temp_est = min_temp + static_cast <float> (rand()) / RAND_MAX * (max_temp - min_temp);
```

Reply:
Resolved in latest commit